### PR TITLE
Filter improvements:

### PR DESCRIPTION
**Pull request by izzytwosheds**
- introduce release method for GL asset cleanup
 - split GL initialization and setting VP matrix

### CODE REVIEW

File: litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java [hunk 1]
```diff
@@ -74,15 +78,27 @@
...
+    @CallSuper
+    public void setVpMatrix(@NonNull float[] vpMatrix, int vpMatrixOffset) {
         // last, we multiply the model matrix by the view matrix to get final MVP matrix for an overlay
         mvpMatrix = GlFilterUtil.createFilterMvpMatrix(vpMatrix, transform);
         mvpMatrixOffset = 0;
     }
 
+    @Override
+    @CallSuper
+    public void release() {
```
Comment (lopeng) [inline litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java:96]: any problem if release() is called before OverlayTexture created or failed?

File: litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java [hunk 2]
```diff
@@ -109,8 +125,17 @@ void renderOverlayTexture(int textureId) {
     int createOverlayTexture(@NonNull Bitmap overlayBitmap) {
         int overlayTextureID;
 
+        vertexShaderHandle = GlRenderUtils.loadShader(GLES20.GL_VERTEX_SHADER, VERTEX_SHADER);
+        if (vertexShaderHandle == 0) {
+            throw new RuntimeException("failed loading vertex shader");
+        }
+        fragmentShaderHandle = GlRenderUtils.loadShader(GLES20.GL_FRAGMENT_SHADER, FRAGMENT_OVERLAY_SHADER);
+        if (fragmentShaderHandle == 0) {
+            throw new RuntimeException("failed loading fragment shader");
```
Comment (lopeng) [inline litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java:134]: vertexShaderHandle will not be free if fragmentShaderHandle failed?

File: litr/src/main/java/com/linkedin/android/litr/filter/GlFilter.java [hunk 3]
```diff
@@ -16,20 +16,32 @@
...
+     * and pass copies of it are passed into to filters using this method, so they can use it to
      * configure their own MVP matrices. Model matrix not present. View matrix is configured to have its camera angle match video's.
      * Projection matrix is configured to orthogonal projection to account for video frame's aspect ratio: (-aspectRatio, aspectRatio, -1, 1, -1, 1)
      * Filters can set up their model matrix and then multiply it by renderer's VP matrix to match their geometry to video frame's.
      *
      * @param vpMatrix VP (projection * view) matrix configured by the renderer, usually
      * @param vpMatrixOffset offset into VP matrix
      */
-    void init(@NonNull float[] vpMatrix, int vpMatrixOffset);
+    void setVpMatrix(@NonNull float[] vpMatrix, int vpMatrixOffset);
```
Comment (lopeng) [inline litr/src/main/java/com/linkedin/android/litr/filter/GlFilter.java:35]: By splitting init() and setVpMatrix(), it enables the client to runtime re-setup VpMatrix. Any potential problem?

File: litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java (hunk 1 above)
Comment (izzytwosheds) [reply litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java:96]: No. `overlayTextureHandle` is initialized to `0`, which `glDeleteTextures` silently ignores: https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glDeleteTextures.xml

File: litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java (hunk 2 above)
Comment (izzytwosheds) [reply litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/BaseOverlayGlFilter.java:134]: Good call. I will `release` the filter on failures, which should all all necessary cleanups.

File: litr/src/main/java/com/linkedin/android/litr/filter/GlFilter.java (hunk 3 above)
Comment (izzytwosheds) [reply litr/src/main/java/com/linkedin/android/litr/filter/GlFilter.java:35]: That is the capability we need for filter preview.

**Review by lopeng — APPROVED** [review]

**Pull request merged**